### ISSUE TITLE
Starting Rails Server fails on Raspberry PI 3 B+

### ISSUE THREAD

**Issue by ultrasam** [ffi/ffi#697]
I´ve been trying to install rbenv, ruby and rails in my Raspberry Pi 3 B+ in the past couple of days and after some issues here and there I've managed to get everything installed with no errors

With rbenv I installed both Ruby 2.6.3 and Ruby 2.5.5
Using Rails 5.2.3

and with both ruby versions I am getting the same problem when I execute the command:
```
rails server
```

### Steps to reproduce

1) On a Raspberry Pi install rbenv using brew
```
brew install rbenv
```

2) Then use the following command to install ruby
```
rbenv install 2.6.3
``` 

3) Then install rails:
```
gem install rails -v 5.2.3
```

4) Create a new folder to hold your application and go to that folder:
```
$ cd                  # Change to the home directory.
$ mkdir environment     # Make a environment directory.
$ cd environment/       # Change into the environment directory.
```

5) Use rails to create the application inside the previous folder:
```
rails _5.2.3_ new hello_app
```

6) Go to the recently created 'hello_app' folder
```
cd hello_app
```
7) Start the rails server
```
rails server
```
### Expected behavior

```
pi@raspberrypi:~/environment/hello_app $ rails server
=> Booting Puma
=> Rails application starting on http://localhost:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
```
And then it should just stay there to be able to access the web app on the address indicated

### Actual behavior

First, the output shows this:
```
pi@raspberrypi:~/environment/hello_app $ rails server
=> Booting Puma
=> Rails 5.2.3 application starting in development 
=> Run `rails server -h` for more startup options
```
And then the contents of the text file in the next link are shown.

Here is a link with the whole output:

https://www.dropbox.com/s/gjgo81zqeekfnvt/railsservererror.txt?dl=0
### System configuration

**Ruby version**: I´ve tried with both 2.6.3 and Ruby 2.5.5
**Rails version**: I've tried with 6.0.0.rc1 and with 5.2.3
**Puma version**: I've tried with 3.12.1 and a previous one too

### Workaround

User @dannyfallon  helped me with a workaround and explanation as follows from the Puma github site because I thought this was an issue with the Puma web server, but I was wrong:

> According to your segfault's `Control frame information` you never even reached Puma. I would think that since Puma was never reached if you swapped `puma` in your gemfile for `thin` or `webrick` you still wouldn't be able to run `rails server`. 
> 
> When you run `rails server` the application does a bunch of stuff before actually passing off the `Rails::Application` to a Rack server to run. In development Rails creates an `EventedFileWatcher` to watch for file changes so it can reload your app without you having to stop/start your server all the time. This uses the `listen` gem, which under the hood uses some OS-specific file events so it doesn't have to keep polling. In trying to create a watch (which `listen` does through `rb-inotify` for Linux), Ruby encountered a segmentation fault:
> 
> ```
> /home/pi/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rb-inotify-0.10.0/lib/rb-inotify/notifier.rb:57: [BUG] Segmentation fault at 0xe92d000e
> ruby 2.5.5p157 (2019-03-15 revision 67260) [armv7l-linux-eabihf]
> 
> -- Control frame information -----------------------------------------------
> c:0095 p:---- s:0527 e:000526 CFUNC  :inotify_init
> c:0094 p:0041 s:0523 e:000522 METHOD /home/pi/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/rb-inotify-0.10.0/lib/rb-inotify/notifier.rb:57 [FINISH]
> ```
> 
> Because `rb-inotify` is using FFI and some C functions, when things go wrong it does tend to blow up badly with segmentation faults.
> 
> There's lots of stuff you can do if you want to go deeper like report a bug over at FFI, attempt to recompile Ruby from source, test `inotify` on your distro outside of Ruby but you probably just want to get on with some development given you've been at it a few days.
> 
> I believe to fix this you can open `config/development.rb` and replace this:
> 
> ```ruby
>   # Use an evented file watcher to asynchronously detect changes in source code,
>   # routes, locales, etc. This feature depends on the listen gem.
>   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
> ```
> 
> with 
> 
> ```ruby
>   # Use a simple file watcher to detect changes in source code
>   config.file_watcher = ActiveSupport::FileUpdateChecker
> ```
> 
> Caveat: The `EventedFileWatcher` is supposed to use OS events to detect file changes. The 
> `FileWatcher` will poll for changes. This may slow down reloads or cause higher system load.

_Originally posted by @dannyfallon in https://github.com/puma/puma/issues/1797#issuecomment-493190148_

The solution proposed by @dannyfallon  actually allowed me to start the Web Server, but I was wondering if the people here at FFI could give me a hand to try to figure out what went wrong and if there is anything I can do to fix this problem and be able to use
```
config.file_watcher = ActiveSupport::EventedFileUpdateChecker
```

Sorry for the long post...

Awaiting your comments.

P.S.: by the way, I get the same segmentation fault when I try to create a new rails app with rails version 6.0.0.rc1:

```
rails _6.0.0.rc1_ new hello_app
```
but if I use version 5.2.3 I don´t get the error

```
rails _5.2.3_ new hello_app
```

Thanks,

**Comment by larskanis** [comment]
Thank you for this well explained report! Still it's very unspecific from the ffi's point of view. The segfault happens somewhere within `inotify_init`, so that I guess that there is an issue in `rb-inotify`.

We regular run ffi on armv7l as part of our CI, and there are no known issues on that platform. Here is an example of such a ffi spec run: https://travis-ci.org/ffi/ffi/jobs/533985370

Could you please file an issue on the rb-inotify repository? It would also help to run the failing script with valgrind and of course to build a minimal failing test example for reproduction.

**Comment by ultrasam** [comment]
I have already issued a report on the rb-inotify repository (https://github.com/guard/rb-inotify/issues/94) but still haven´t got an answer.
I issued this report here as well because it was recommended by the guy who helped me with the workaround.
You may be totally right about this potentially be a problem of `rb-inotify` I just wanted to inform you guys and maybe you could have an idea of what was going on.
Can you guide me through the process of running the failing script with valgrind? I am quite a newbie with Linux so I mostly try to follow tutorials and fix problems via forums.

I would really appreciate any assistance given.

thanks,

**Comment by dannyfallon** [comment]
:wave: Sorry about this @ultrasam, I didn't fully understand what FFI was doing. I no longer think it's a bug with FFI but something in `rb-inotify`, your Ruby or your distro 😓 . Before going and trying to use Valgrind I'd see what the folks over in rb-inotify have to say :)

**Comment by ultrasam** [comment]
Perfect, thanks @dannyfallon , I really want to get to the bottom of this. It is driving me crazy

**Comment by utherpendragon** [comment]
I have had a similar problem on  a raspberry pi 2b after a gem upgrade. Running bin/rspec using the spring gem caused an almost identical error. On backing out and upgrading gems one at a time with rails 5.2.3 + ruby 2.6.4 worked as long as I  forced ffi to version 1.9.18.

Given this working set of gems I attempted to upgrade to ffi 1.9.21 and it failed to build the native extensions. The gem install hint given by the bundle upgrade failure also failed with a complaint about not finding ffi.h!!!
I have looked through the ffi documentation and found references to libffi-dev package.
I installed this raspbian package and reinstalled ffi 1.9.21 via 'gem install' which worked.  bin/rspec also worked.
I have now successfully reinstalled ffi 1.11.1 and upgraded the rails app.
I believe that the problem boils down to ffi on Linux on at least  an arm CPU ffi later than 1.9.18 assumes the presence of the ffi development package.  There are no warnings in the install process and as ffi is only installed as a dependency, the end user would not be expecting to do anything specifically to make any requested gem work.

I suggest some sort of check in the install process to warn if the ffi development library is not present and what to do if it isn't.  This would have saved me hours of painstakingly updating 1 gem at a time to isolate the problem.

I suggest that the original issue is likely to be solved by installing libffi-dev on the RPI and running 'gem install ffi -v '<your version here>' to rebuild the native extension.
